File: solutions/2023/pughmds/ruby/day05/main.rb
require 'set'
# First attempt: worked great for part 1, but too slow for part 2 (it'd take 3.5 days to get an answer on my
# really old personal computer). This is a brute-force method. It'd get the job done, but take forever. There
# are faster ways to do this, such as using the mapping functions.

def parse_maps(maps)
  map_pipe = {}

  maps.each_with_index do |map_values, i|
    values = map_values.split("\n")[1..]
    map_pipe[i] ||= []

    values.each do |map_val|
      next if map_val.empty?

      input_data = map_val.split.map(&:to_i)
      drs, srs, rl = input_data
      map_pipe[i] << { "dest_range_start" => drs, "source_range_start" => srs, "range_length" => rl }
    end
  end

  map_pipe
end

def parse_input(data)
  instructions_text = data.join("\n")
  sections = instructions_text.split(/\n\s*\n/)
  seeds = sections[0].split(":")[1].strip.split.map(&:to_i)
  processes = []

  sections[1..].each do |sect|
    process = []
    seed_data = sect.split(":")[1].strip.split("\n")

    seed_data.each do |item|
      seed_strings = item.strip.split.map(&:to_i)
      process << seed_strings
    end

    processes << process
  end

  [seeds, processes]
end

def part_one(data)
  seeds, processes = parse_input(data)
  seeds = Set.new(seeds)
  result = []

  seeds.each do |seed|
    processes.each do |step|
      step.each do |process|
        diff = process[0] - process[1]

        if (process[1]..(process[1] + process[2] - 1)).cover?(seed)
          seed += diff
          break
        end
      end
    end

    result << seed
    puts result
  end

  result.min
end

def part_two(data)
  seeds, processes = parse_input(data)
  seeds = seeds.each_slice(2).to_a
  min_loc = Float::INFINITY

  seeds.each do |seed_group|
    start_seed, range_length = seed_group
    puts start_seed
    (start_seed..(range_length + start_seed - 1)).each do |i|
      r = i

      processes.each do |step|
        step.each do |process|
          diff = process[0] - process[1]

          if (process[1]..(process[1] + process[2] - 1)).cover?(r)
            r += diff
            break
          end
        end
      end
      min_loc = [min_loc, r].min
    end
  end

  min_loc
end






# Specify the file path
file_path = '../../inputs/day05/input.txt'

# Initialize an empty array to store data
data = []

# Open the file and read its contents into the array
File.open(file_path, 'r') do |file|
  file.each_line do |line|
    data << line.chomp
  end
end

# Run part 1
#result = part_one(data)
#print "The Part 1 Answer is: ", result, "\n"

# Run part 2
result = part_two(data)
print "The Part 2 Answer is: ", result, "\n"

File: solutions/2023/pughmds/python/day05/__main__.py
# First attempt: worked great for part 1, but too slow for part 2 (it'd take 3.5 days to get an answer on my
# really old personal computer). This is a brute-force method. It'd get the job done, but take forever. There
# are faster ways to do this, such as using the mapping functions.

EXPECTED_TEST_ANSWER_PART1 = [35]
EXPECTED_TEST_ANSWER_PART2 = [46]

import re
import numpy as np


def parse_maps(maps):
    map_pipe = {}

    for i, map_values in enumerate(maps):
        values = map_values.split("\n")[1:]
        for map_val in values:
            if not map_val:
                continue
            input_data = list(map(int, map_val.split()))
            drs, srs, rl = input_data
            if i not in map_pipe:
                map_pipe[i] = []
            map_pipe[i].append(
                {"dest_range_start": drs, "source_range_start": srs, "range_length": rl}
            )

    return map_pipe


def parseInput(data):
    # This is one of those programs that would have been easier to handle by taking the entire input
    # file as one string and splitting from there. Because I don't want to rewrite my shared input script
    # We can just combine it back here:
    instructions_text = "\n".join(data)

    # This should roughly split the text into sections
    sections = re.split(r"\n\s*\n", instructions_text)
    seeds = sections[0].split(":")[1].strip().split(" ")
    processes = []
    for idx, sect in enumerate(sections):
        if idx == 0:
            continue
        process = []
        seed_data = sect.split(":")[1].strip().split("\n")
        for item in seed_data:
            seed_strings = item.strip().split(" ")
            process.append([int(s) for s in seed_strings])
        processes.append(process)

    return seeds, processes


def run(data):

    # Determines if the numbers on the right appear in the winning number list on the left
    # Scores based on a power of two depending on how many matches there are.

    seeds, processes = parseInput(data)
    seeds = set([int(s) for s in seeds])
    result = []
    for seed in seeds:
        for step in processes:
            for process in step:
                diff = process[0] - process[1]
                if process[1] <= seed <= (process[1] + process[2] - 1):
                    seed = seed + diff
                    break
        result.append(seed)
        print(result)

    return min(result)


def run_p2(data):
    '''
        Same as above, but crazy rules involving...more scratch cards?
        Takes a LONG time to run, so use with caution. Probably could
        speed up greatly with some tricks in mapping, but I have my
        answer and need to catch up so will save for later.
    '''
    seeds, processes = parseInput(data)
    seeds = [int(s) for s in seeds]

    # Pair the seed values up...
    min_loc = 999999999
    for seed_group in zip(seeds[::2], seeds[1::2]):
        print(seed_group)
        start_seed = seed_group[0]
        range_length = seed_group[1]
        for i in range(start_seed, range_length + start_seed - 1):
            r = i
            for step in processes:
                for process in step:
                    diff = process[0] - process[1]
                    if process[1] <= r <= (process[1] + process[2] - 1):
                        r = r + diff
                        break
            min_loc = min(min_loc, r)

    return min_loc
